### PR TITLE
[SPARK-24237][SS] Continuous shuffle dependency and map output tracker

### PR DESCRIPTION
**Pull request by xuanyuanking**
## What changes were proposed in this pull request?

As our disscussion in [jira comment](https://issues.apache.org/jira/browse/SPARK-24036?focusedCommentId=16470067&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-16470067) and [design comment](https://docs.google.com/document/d/1IL4kJoKrZWeyIhklKUJqsW-yEN7V7aL05MmM65AYOfE/edit?disco=AAAAB4X1H_E) and [design doc](https://docs.google.com/document/d/14cGJ75v9myznywtB35ytEqL9wHy9xfZRv06B6g2tUgI/edit#bookmark=id.2lfv2glj7ny0), this pr including the following changes:
1. Add ContinuousShuffleDependency support, which can allow shuffleId generated from epoch and re-register shuffleHandle with new shuffleId.
2. Add ContinuousMapOutputTrackerWorker, which can get shuffle status by a blocking way, and support register shuffle\map ouput in tracker worker.
3. Add ContinuousShuffleMapTask.

## How was this patch tested?

Add a new UT for ContinuousMapOutputTracker.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/Dependency.scala [hunk 1]
```diff
@@ -65,15 +65,17 @@ abstract class NarrowDependency[T](_rdd: RDD[T]) extends Dependency[T] {
  * @param keyOrdering key ordering for RDD's shuffles
  * @param aggregator map/reduce-side aggregator for RDD's shuffle
  * @param mapSideCombine whether to perform partial aggregation (also known as map-side combine)
+ * @param isContinuous mark the dependency is base for continuous processing or not
```
Comment (jose-torres) [inline core/src/main/scala/org/apache/spark/Dependency.scala]: If we need a bit to mark whether it's continuous or not, ContinuousShuffleDependency shouldn't be inheriting ShuffleDependency in the first place.

File: core/src/main/scala/org/apache/spark/Dependency.scala [hunk 2]
```diff
@@ -88,14 +90,53 @@ class ShuffleDependency[K: ClassTag, V: ClassTag, C: ClassTag](
...
+    rdd,
+    dep.partitioner,
+    dep.serializer,
+    dep.keyOrdering,
+    dep.aggregator,
+    dep.mapSideCombine, true) {
+
+  val baseShuffleId: Int = dep.shuffleId
+
+  override val shuffleId: Int = continuousEpoch * totalShuffleNum + baseShuffleId
```
Comment (jose-torres) [inline core/src/main/scala/org/apache/spark/Dependency.scala]: We surely can't artificially construct shuffle IDs like this. IIUC only shuffle IDs returned from SparkContext.newShuffleId() are valid shuffle IDs.

File: core/src/main/scala/org/apache/spark/MapOutputTracker.scala [hunk 3]
```diff
@@ -233,6 +239,28 @@ private[spark] class MapOutputTrackerMasterEndpoint(
...
+      logInfo(s"Checking missing partitions for $shuffleId")
+      // If get None from findMissingPartitions, just return a non-empty Seq
+      val missing = tracker.findMissingPartitions(shuffleId).getOrElse(Seq(0))
+      if (missing.isEmpty) {
+        context.reply(true)
+      } else {
+        context.reply(false)
+      }
+
+    case CheckAndRegisterShuffle(shuffleId: Int, numMaps: Int) =>
```
Comment (jose-torres) [inline core/src/main/scala/org/apache/spark/MapOutputTracker.scala]: There's already a way to register shuffles. Why is this RPC needed?

File: core/src/main/scala/org/apache/spark/MapOutputTracker.scala [hunk 4]
```diff
@@ -233,6 +239,28 @@ private[spark] class MapOutputTrackerMasterEndpoint(
...
+      logInfo(s"Trying to register shuffle $shuffleId")
+      if (!tracker.shuffleStatuses.contains(shuffleId)) {
+        tracker.registerShuffle(shuffleId, numMaps)
+        logDebug(s"Shuffle $shuffleId doesn't exist, register it now")
+      }
+      context.reply(true)
+
+    case RegisterMapOutput(shuffleId: Int, mapId: Int, status: MapStatus) =>
```
Comment (jose-torres) [inline core/src/main/scala/org/apache/spark/MapOutputTracker.scala]: same question as above

File: core/src/main/scala/org/apache/spark/SparkEnv.scala [hunk 5]
```diff
@@ -227,6 +228,7 @@ object SparkEnv extends Logging {
       mockOutputCommitCoordinator: Option[OutputCommitCoordinator] = None): SparkEnv = {
 
     val isDriver = executorId == SparkContext.DRIVER_IDENTIFIER
+    val isContinuous = conf.getBoolean("spark.streaming.continuousMode", false)
```
Comment (jose-torres) [inline core/src/main/scala/org/apache/spark/SparkEnv.scala]: If I don't pass the right option at startup, I can never run a continuous processing query? I don't think that's the right model.

File: core/src/main/scala/org/apache/spark/scheduler/ContinuousShuffleMapTask.scala [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+    } else 0L
+
+    var writer: ShuffleWriter[Any, Any] = null
+    val manager = SparkEnv.get.shuffleManager
+    val mapOutputTracker = SparkEnv.get.mapOutputTracker
+      .asInstanceOf[ContinuousMapOutputTrackerWorker]
+
+    while (!context.isCompleted() || !context.isInterrupted()) {
+      try {
+        // Create a ContinuousShuffleDependency which has new shuffleId based on continuous epoch.
```
Comment (jose-torres) [inline core/src/main/scala/org/apache/spark/scheduler/ContinuousShuffleMapTask.scala]: Who's cleaning up the old one?

File: core/src/main/scala/org/apache/spark/MapOutputTracker.scala [hunk 7]
```diff
@@ -769,6 +796,43 @@ private[spark] class MapOutputTrackerWorker(conf: SparkConf) extends MapOutputTr
   }
 }
 
+/**
+ * MapOutputTrackerWorker for continuous processing, the main difference with MapOutputTracker
+ * is waiting for a time when the upstream's map output status not ready.
```
Comment (jose-torres) [inline core/src/main/scala/org/apache/spark/MapOutputTracker.scala]: We can't change the behavior for queries which aren't continuous processing.

File: core/src/main/scala/org/apache/spark/scheduler/ContinuousShuffleMapTask.scala [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+      threadMXBean.getCurrentThreadCpuTime
+    } else 0L
+    val ser = SparkEnv.get.closureSerializer.newInstance()
+    // TODO: rdd here should be a wrap of ShuffledRowRDD which never stop
+    val (rdd, dep) = ser.deserialize[(RDD[_], ShuffleDependency[_, _, _])](
+      ByteBuffer.wrap(taskBinary.value), Thread.currentThread.getContextClassLoader)
+    _executorDeserializeTime = System.currentTimeMillis() - deserializeStartTime
+    _executorDeserializeCpuTime = if (threadMXBean.isCurrentThreadCpuTimeSupported) {
+      threadMXBean.getCurrentThreadCpuTime - deserializeStartCpuTime
+    } else 0L
```
Comment (jose-torres) [inline core/src/main/scala/org/apache/spark/scheduler/ContinuousShuffleMapTask.scala]: There's a lot of code here that has nothing to do with continuous processing. I'm not sure how we can be confident it's correct and will remain correct as the scheduler evolves.

**Review by jose-torres — COMMENTED** [review]
In general, I have the same comment I mentioned in SPARK-24036. This PR tightly integrates with the internals of the scheduler in unsafe ways, and I don't see what benefit we're getting from doing that rather than writing RDDs on top.

File: core/src/main/scala/org/apache/spark/Dependency.scala (hunk 1 above)
Comment (xuanyuanking) [reply core/src/main/scala/org/apache/spark/Dependency.scala]: Actually we can implement ContinuousShuffleDependency by inheriting Dependency. In current way few interface can be changed.

File: core/src/main/scala/org/apache/spark/scheduler/ContinuousShuffleMapTask.scala (hunk 6 above)
Comment (xuanyuanking) [reply core/src/main/scala/org/apache/spark/scheduler/ContinuousShuffleMapTask.scala]: The old one(ShuffleDependency) register cleaner on driver side. ContinuousShuffleDependency we created on executor side, the tmp data delete by epoch coordinator receive commit event(Not in this PR).

File: core/src/main/scala/org/apache/spark/MapOutputTracker.scala (hunk 7 above)
Comment (xuanyuanking) [reply core/src/main/scala/org/apache/spark/MapOutputTracker.scala]: Maybe we didn't change here, IIUC in batch mode, child stage only submitted while no missing partition of parent stage.

File: core/src/main/scala/org/apache/spark/Dependency.scala (hunk 1 above)
Comment (jose-torres) [reply core/src/main/scala/org/apache/spark/Dependency.scala]: Considering ContinuousShuffleDependency to be a special kind of ShuffleDependency doesn't make sense if we need to add this much special casing to do it.

File: core/src/main/scala/org/apache/spark/Dependency.scala (hunk 2 above)
Comment (jose-torres) [reply core/src/main/scala/org/apache/spark/Dependency.scala]: Sorry, I'm not sure I understand. The SparkContext is responsible for assigning shuffle IDs, so it's not valid to register a ContinuousShuffleDependency with a shuffle ID which SparkContext did not assign.

File: core/src/main/scala/org/apache/spark/MapOutputTracker.scala (hunk 7 above)
Comment (jose-torres) [reply core/src/main/scala/org/apache/spark/MapOutputTracker.scala]: We can't really make changes that maybe don't change existing behavior. They need to definitely not change existing behavior, ideally by not using the same codepath if they need to do different things.

File: core/src/main/scala/org/apache/spark/scheduler/ContinuousShuffleMapTask.scala (hunk 6 above)
Comment (jose-torres) [reply core/src/main/scala/org/apache/spark/scheduler/ContinuousShuffleMapTask.scala]: The cleanup should probably be in the same PR as the registration.

Won't that strategy leak resources if the query ends up shut down before the commit event is received?

File: core/src/main/scala/org/apache/spark/Dependency.scala (hunk 1 above)
Comment (xuanyuanking) [reply core/src/main/scala/org/apache/spark/Dependency.scala]: Got it, if we think currently implementation is tricky here, we can change the implementation by getting rid of the ContinuousShuffleDependency, just as you said in jira comments :"We might not need this to be an actual org.apache.spark.Dependency."

File: core/src/main/scala/org/apache/spark/Dependency.scala (hunk 2 above)
Comment (xuanyuanking) [reply core/src/main/scala/org/apache/spark/Dependency.scala]: Got it, if we move EpochCoordinator to SparkEnv, I think we can re-implement the shuffle register on driver side, totally controlled by EpochCoordinator. Even if the EpochCoordinator doesn't manage the shuffle register work, I think EpochCoordinator should move to SparkEnv and take more work in shuffle support in your design, am I right?

File: core/src/main/scala/org/apache/spark/MapOutputTracker.scala (hunk 7 above)
Comment (xuanyuanking) [reply core/src/main/scala/org/apache/spark/MapOutputTracker.scala]: The current implement of ContinuousShuffleMapOutputTracker didn't change existing behavior, just choose the corresponding output tracker instance by the job type. After we use a better way of mark a job is continuous mode(discuss in https://github.com/apache/spark/pull/21293#discussion-diff-187368178R231), they do different things and not use same codepath.

File: core/src/main/scala/org/apache/spark/scheduler/ContinuousShuffleMapTask.scala (hunk 8 above)
Comment (xuanyuanking) [reply core/src/main/scala/org/apache/spark/scheduler/ContinuousShuffleMapTask.scala]: Currently we have passed some shuffle demo over CP mode by this way, of cause we need more check on both correctness of both batch and cp job.

**Comment by xuanyuanking** [comment]
@jose-torres Great thanks for you advise and guidance for us! I found the main difference between us is whether we can reuse current implementation of scheduler and shuffle. I marked in your design here:https://docs.google.com/document/d/1IL4kJoKrZWeyIhklKUJqsW-yEN7V7aL05MmM65AYOfE/edit?disco=AAAAB4z058g

**Comment by gaborgsomogyi** [comment]
@xuanyuanking are you still working on this? If not may I ask to close?

**Comment by xuanyuanking** [comment]
Yeah, I'll close all the outdated preview PR.

**Pull request closed without merging**